xpra/server/auth/keycloak_auth.py

```diff
@@ -0,0 +1,125 @@
+        return "keycloak"
+
+    def get_challenge(self, digests):
+        assert not self.challenge_sent
+        if "keycloak" not in digests:
+            log.error("Error: client does not support keycloak authentication")
+            return None
+        self.challenge_sent = True
+        return self.salt, "keycloak"
+
+    def check(self, response_json) -> bool:
+        assert self.challenge_sent
+        #log("response_json: %r", repr(response_json))
+        
+        if not response_json:
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False
+        
+        try:
+          response = json.loads(response_json)
+        except Exception:
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False 
+
+        if not isinstance(response, dict):
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False
+
+        log("check(%r)", repr(response))
+
+        auth_code = response.get("code")
+        error = response.get("error")
+
+        if not error and not auth_code:
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False
+
+        if error:
+          log.error("Error: keycloak authentication failed with %s: %s", error, response.get("error_description"))
+          return False
+
+        if auth_code:
+          # Configure client
+          keycloak_openid = KeycloakOpenID(server_url=KEYCLOAK_SERVER_URL,
+                            client_id=KEYCLOAK_CLIENT_ID,
+                            realm_name=KEYCLOAK_REALM_NAME,
+                            client_secret_key=KEYCLOAK_CLIENT_SECRET_KEY)
+
+          try:
+            # Get well_known
+            config_well_know = keycloak_openid.well_know()
+            log("well_known: %r", repr(config_well_know))
+        
+            # Get token
+            token = keycloak_openid.token(code=auth_code, grant_type=[KEYCLOAK_GRANT_TYPE], redirect_uri=KEYCLOAK_REDIRECT_URI)
+
+            # Verify token
+            token_info = keycloak_openid.introspect(token['access_token'])
+            log("token_info: %r", repr(token_info))
+
+            if token_info["active"]:
+              # Get userinfo
+              user_info = keycloak_openid.userinfo(token['access_token'])
+              log("userinfo_info: %r", repr(user_info))
+
+              log("keycloak authentication succeeded: token is active")
+              return True
+            else:
+              log.error("Error: keycloak authentication failed as token is not active")
+              return False
+          except Exception as e:
+            log.error("Error: keycloak authentication failed with code %s: %s", e.response_code, e.error_message)
```

Comment:
You are catching a generic `Exception` type so you cannot access `response_code` and `error_message` on it.
Catch a more specific type that is known to have these attributes. The generic `Exception` type should be caught by the caller.